lisp/ess-custom.el

```diff
@@ -2782,21 +2790,52 @@ default or not."
 
 
 ;;; fl-keywords R
-(defvar ess-R-fl-keyword:modifiers
-  (cons (concat "\\(" (regexp-opt ess-R-modifyiers 'words) "\\)\\s-*(")
-        '(1 ess-modifiers-face))     ; modify search list or source (i.e. directives)
-  "Font-lock keyword R modifiers.")
+(defvar ess-R-fl-keyword:modifiers nil
+  "Font-lock regexp for search path modifiers.")
+
+(defvar ess-R-fl-keyword:fun-defs nil
+  "Font-lock regexp for function definitions.")
+
+(defvar ess-R-fl-keyword:bare-keywords nil
+  "Font-lock regexp for keywords that do not precede an opening parenthesis.")
+
+(defvar ess-R-fl-keyword:keywords nil
+  "Font-lock regexp for keywords that precede an opening parenthesis.")
+
+(defvar ess-R-fl-keyword:control-flow-keywords nil
+  "Font-lock regexp for control flow keywords.")
 
-(defvar ess-R-fl-keyword:fun-defs
-  (cons ess-R-function-name-regexp
-        '(1 font-lock-function-name-face nil))
-  "Font-lock keyword - function defintions for R.")
+(defvar ess-R-fl-keyword:signal-keywords nil
+  "Font-lock regexp for signal keywords.")
+
+(defvar ess-R-fl-keyword:assign-ops nil
```

Comment:
I really dislike that all these variables are initialized with `nil` and then set by side effect elsewhere. Cannot we do better here?

Reply:
How would you do it without code duplication? A defvar is side-effectful as well. Or are you concerned about the semantics of initialisation?

Reply:
If the latter, we could have an `&optional init` argument to signal we shouldn't override variables that are already set.

Reply:
I was concerned mostly with the stylistic part of it. When I navigate to the var definition and I see nil, it's disturbing. I have to go and figure where and how it is initialized.

The initialization issue is even more imporant. If people "customized" those variables we will be overwriting them.

Reply:
Is the sole motivation to reduce the burden of restarting emacs after setting those vars? If so, I don't see the encured complications worth the tradeoff.

Reply:
Plus add to this that we are introducing one extra function and making both users and developer read and understand it.  With all those questions naturally raised "Why? How? Why not standard?". It's complicated. We are in cutting and cleaning season,  not adding extra complexities.

There are probably other variables in ESS which cannot be easily re-initialized.  I don't think we should introduce an init function for each of those.

Reply:
> Is the sole motivation to reduce the burden of restarting emacs after setting those vars? If so, I don't see the encured complications worth the tradeoff.

No, it's to make it possible to _modify_ the default list of keywords without having to copy the default list (and freeze it for good, missing out on upstream changes) in the user config.

There is currently no way to modify these. This function makes things simpler, not more complicated.

Reply:
I don't think `font-lock-remove-keywords` is practical. IIUC you have to provide the exact keyword spec right? But this spec has been constructed by ESS and is not trivially reproducible. So while it's easy to add new keywords, it's hard to remove them. Letting users customise `-keywords` variables is a good feature from that point of view.

Reply:
> Also I bet we leave many users perplexed as to why their modification of the keywords variables (after having loaded ESS) do not have any effect. 

That's pretty standard and I thinks users are used to that. 

> missing out on upstream changes

Well, that is the case with all customized defcustoms and all `setq`ed variables. I wouldn't put too much weight on that. Your proposal doesn't really help with that. If a user already have them customized, our new changes won't affect them.

Actually, in the next version, we can issue a warning if we detect that someone had those keywords set. We might also go for some dynamic version. If some packages do provide special set of keywords, we can pick them up automatically. CIDER has such a feature for macros. Or maybe a special set `ess-user-custom-keywords` is warranted here.

I do agree with @jabranham here. User's shouldn't ideally be customizing those variables, and if they do it's sort of their problem to solve. Plus, they should use `font-lock-add-keywords` for new stuff. Why users would need to remove keywords?

Reply:
> That's pretty standard and I thinks users are used to that.

They may be used to being perplexed and frustrated (I know I was, when learning ESS) but that's not a good state of affairs. We should make it easy to configure ESS and clearly document things that are not meant to be configured and the alternative process. `ess-R-keywords`-like variables are currently deceptive.

> Why would users remove keywords?

Because they don't like them? The main use case is to add to the keywords though, without having to worry about regexps (matching the open paren etc).

If the only thing supported user interaction for those keywords is to set them to nil, this should be documented. And we should move all these variables to `ess-r--` if they are internal.

Overall I don't see it a big deal to have the keywords variables as part of the configuration UI. And it is standard to have a follow-up function to execute after setting a variable (which is automatically run by the customisation system when the defvar is promoted to defcustom).

In any case we should be clear about what's part of the UI and what's private.

Reply:
> that is the case with all customized defcustoms and all setqed variables.

That is not true, you can usually modify variables and customs.

Reply:
>>    that is the case with all customized defcustoms and all setqed variables.

> That is not true, you can usually modify variables and customs.

Don't know what is so much "not true" here. If a user customizes a value your change upstream will not be propagated into the user config. It's a null argument.

Reply:
You can usually both read from and write to a variable, which is currently not the case with the keywords variables. What is a null argument?

Reply:
We are talking about different things. I was referring to your comment that they will miss out on upstream change - "having to copy the default list (and freeze it for good, missing out on upstream changes) in the user config." . You focus on how within session changes propagate into the application.  Anyways, let's move into the main thread and let's think how we can make changes propagate without adding any new  "generators" and code duplication.

Reply:
There is currently no way to read from the default keyword lists and write a new list. You can either read from it after loading ESS, or write to it before loading ESS. Doing both is not possible, not within session, and not even after restarting Emacs. Hence there is no way to pick up the default upstream list and append to it, which I don't think is an unusual case (for example there are packages out there that provide additional control-flow functions).

I agree it is preferable for the changes to be automatically picked up, I'll think about your suggestion of introducing laziness, thanks.

Reply:
Ok. I see. "upstream" usually means the code in the main repo out there. So I thought you differentiate between "upstream" (our settings) and user settings. 

> Hence there is no way to pick up 

Yes. I think we all agree this is a problem. We disagree how to deal with it. 

It makes me wonder if we have similar "hardwiring" elsewhere in ESS. I think we do, but I am too busy right now to dwell and look for those. In any case the font-lock is probably the most important case.